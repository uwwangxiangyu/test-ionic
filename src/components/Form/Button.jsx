import React from 'react'
import { useFormikContext } from 'formik'
import { IonButton } from '@ionic/react'

const FormButton = ({ title }) => {
  const { handleSubmit } = useFormikContext()

  return <IonButton color="primary" expand="block" onClick={() => handleSubmit()}>{title}</IonButton>
}

export default FormButton
