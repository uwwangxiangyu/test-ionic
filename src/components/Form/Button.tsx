import React, {FC} from 'react'
import {useFormikContext} from 'formik'
import {IonButton} from "@ionic/react";

export interface ButtonProps {
    title: string;
}

const FormButton: FC<ButtonProps> = ({title}) => {
    const {handleSubmit} = useFormikContext()

    return <IonButton color="primary" expand="block" onClick={() => handleSubmit()}>{title}</IonButton>
}

export default FormButton
