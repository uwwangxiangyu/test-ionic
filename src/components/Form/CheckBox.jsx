import React, { useMemo } from 'react'
import { IonCheckbox, IonItem, IonLabel } from '@ionic/react'
import { useFormikContext } from 'formik'

const CheckBox = ({ name, label }) => {
  const {
    setFieldValue,
    values
  } = useFormikContext()

  const valueSet = useMemo(() => new Set(values[name]), [values[name]])

  return (
    <IonItem>
      {label && (<IonLabel>{label}</IonLabel>)}
      <IonCheckbox
        onIonChange={() => {
          if (valueSet.has(label)) {
            valueSet.delete(label)
          } else {
            valueSet.add(label)
          }
          setFieldValue(name, Array.from(valueSet))
        }}
        checked={valueSet.has(label)}
        slot="start"
      />
    </IonItem>
  )
}

export default CheckBox