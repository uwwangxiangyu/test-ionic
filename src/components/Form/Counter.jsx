import React from 'react'
import { IonIcon, IonInput, IonItem, IonLabel } from '@ionic/react'
import { useFormikContext } from 'formik'
import { addCircleOutline, removeCircleOutline } from 'ionicons/icons'

const Counter = ({ name, label }) => {
  const {
    setFieldValue,
    values
  } = useFormikContext()

  return (
    <IonItem>
      <IonLabel>{label}</IonLabel>
      <IonItem>
        <IonInput type={'number'} value={values[name]}/>
        <IonIcon icon={removeCircleOutline} onClick={() => setFieldValue(name, Math.max(values[name] - 1, 0))}
                 slot={'start'} title={'counterIncrement'}/>
        <IonIcon icon={addCircleOutline} onClick={() => setFieldValue(name, Math.max(values[name] + 1, 0))}
                 slot={'end'} title={'counterDecrement'}/>
      </IonItem>
    </IonItem>
  )
}

export default Counter