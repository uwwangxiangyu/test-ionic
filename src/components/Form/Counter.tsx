import React, {FC} from 'react'
import {IonIcon, IonInput, IonItem, IonLabel} from '@ionic/react'
import {useFormikContext} from "formik";
import {addCircleOutline, removeCircleOutline} from "ionicons/icons";

export interface CounterProps {
    name: string;
    label?: string;
}

const Counter: FC<CounterProps> = ({name, label}) => {
    const {
        setFieldValue,
        values,
    } = useFormikContext<any>()

    return (
        <IonItem>
            <IonLabel>{label}</IonLabel>
            <IonItem>
                <IonInput type={'number'} value={values[name]}/>
                <IonIcon icon={removeCircleOutline} onClick={() => setFieldValue(name, Math.max(values[name] - 1, 0))}
                         slot={'start'} title={'counterIncrement'}/>
                <IonIcon icon={addCircleOutline} onClick={() => setFieldValue(name, Math.max(values[name] + 1, 0))}
                         slot={'end'} title={'counterDecrement'}/>
            </IonItem>
        </IonItem>
    );
};

export default Counter;