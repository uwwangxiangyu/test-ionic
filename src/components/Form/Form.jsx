import React from 'react'
import { Formik } from 'formik'

const Form = ({ children, initialValues, onSubmit, innerRef }) => {
  return (
    <Formik
      validateOnBlur={false}
      initialValues={initialValues}
      onSubmit={onSubmit}
      innerRef={innerRef}
    >
      <div style={{ width: '100%' }}>
        {children}
      </div>
    </Formik>
  )
}

export default Form