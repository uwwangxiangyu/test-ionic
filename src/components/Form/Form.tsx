import React, {FC} from 'react'
import {Formik} from 'formik'

export interface InputProps {
    initialValues: any;
    onSubmit: any;
    innerRef?: any;
}

const Form: FC<InputProps> = ({children, initialValues, onSubmit, innerRef}) => {
    return (
        <Formik
            validateOnBlur={false}
            initialValues={initialValues}
            onSubmit={onSubmit}
            innerRef={innerRef}
        >
            <div>
                {children}
            </div>
        </Formik>
    )
}

export default Form