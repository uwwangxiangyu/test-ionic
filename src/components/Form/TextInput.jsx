import React from 'react'
import { IonInput, IonItem, IonLabel } from '@ionic/react'
import { useFormikContext } from 'formik'

const TextInput = ({ name, label, type }) => {
  const {
    setFieldValue,
    values
  } = useFormikContext()

  return (
    <IonItem style={{ width: '100%' }}>
      {label && (<IonLabel position="floating">{label}</IonLabel>)}
      <IonInput
        onIonChange={(e) => setFieldValue(name, e.detail.value)}
        value={values[name]}
        title={'ionInput'}
        type={type}
        style={{ width: '100%' }}
      />
    </IonItem>
  )
}

export default TextInput