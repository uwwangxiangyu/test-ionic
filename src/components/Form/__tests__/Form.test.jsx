import { render, waitFor } from '@testing-library/react'
import { ionFireEvent as fireEvent } from '@ionic/react-test-utils'
import TextInput from '../TextInput'
import FormButton from '../Button'
import Form from '../Form'
import React from 'react'
import Counter from '../Counter'

test('Form input', async () => {
  const submit = jest.fn()
  const { findByTitle, findByText, container } = render(
    <Form initialValues={{ Fruit: '' }} onSubmit={submit}>
      <TextInput name={'Fruit'} label={'Fruit'}/>
      <FormButton title={'submit'}/>
    </Form>
  )

  const input = await findByTitle('ionInput')
  const button = await findByText('submit')
  fireEvent.ionChange(input, 'apple')
  fireEvent.click(button)

  await waitFor(() => expect(submit.mock.calls[0][0]).toEqual({ Fruit: 'apple' }))
  expect(container).toMatchSnapshot()
})

test('Form counter', async () => {
  const submit = jest.fn()
  const { findByTitle, findByText, container } = render(
    <Form initialValues={{ CampaignPhoto: 0 }} onSubmit={submit}>
      <Counter name={'CampaignPhoto'} label={'Campaign Photo'}/>
      <FormButton title={'submit'}/>
    </Form>
  )

  const increment = await findByTitle('counterIncrement')
  const decrement = await findByTitle('counterDecrement')
  const button = await findByText('submit')
  fireEvent.click(increment)
  fireEvent.click(increment)
  fireEvent.click(decrement)
  fireEvent.click(button)

  await waitFor(() => expect(submit.mock.calls[0][0]).toEqual({ CampaignPhoto: 1 }))
  expect(container).toMatchSnapshot()
})