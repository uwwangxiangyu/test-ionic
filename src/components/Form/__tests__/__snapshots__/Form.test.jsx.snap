// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form counter 1`] = `
<div>
  <div
    style="width: 100%;"
  >
    <ion-item>
      <ion-label>
        Campaign Photo
      </ion-label>
      <ion-item>
        <ion-input
          type="number"
          value="1"
        />
        <ion-icon
          icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Remove Circle</title><path d='M448 256c0-106-86-192-192-192S64 150 64 256s86 192 192 192 192-86 192-192z' stroke-miterlimit='10' class='ionicon-fill-none ionicon-stroke-width'/><path stroke-linecap='round' stroke-linejoin='round' d='M336 256H176' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          slot="start"
          title="counterIncrement"
        />
        <ion-icon
          icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Add Circle</title><path d='M448 256c0-106-86-192-192-192S64 150 64 256s86 192 192 192 192-86 192-192z' stroke-miterlimit='10' class='ionicon-fill-none ionicon-stroke-width'/><path stroke-linecap='round' stroke-linejoin='round' d='M256 176v160M336 256H176' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          slot="end"
          title="counterDecrement"
        />
      </ion-item>
    </ion-item>
    <ion-button
      color="primary"
      expand="block"
    >
      submit
    </ion-button>
  </div>
</div>
`;

exports[`Form input 1`] = `
<div>
  <div
    style="width: 100%;"
  >
    <ion-item
      style="width: 100%;"
    >
      <ion-label
        position="floating"
      >
        Fruit
      </ion-label>
      <ion-input
        style="width: 100%;"
        title="ionInput"
        value="apple"
      />
    </ion-item>
    <ion-button
      color="primary"
      expand="block"
    >
      submit
    </ion-button>
  </div>
</div>
`;
